layer/decorators/definitions.py

```diff
@@ -0,0 +1,164 @@
+import hashlib
+import inspect
+import os
+import pickle  # nosec import_pickle
+import shutil
+import uuid
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Any, Callable, Iterator, List, Optional
+
+import cloudpickle  # type: ignore
+
+from layer.config import DEFAULT_FUNC_PATH
+from layer.contracts.assertions import Assertion
+from layer.contracts.asset import AssetPath, AssetType
+from layer.contracts.fabrics import Fabric
+from layer.contracts.project_full_name import ProjectFullName
+
+
+@dataclass(frozen=True)
+class ResourcePath:
+    # Local file system path of the resource (file or dir), relative to the project dir.
+    # Examples: data/test.csv, users.parquet
+    path: str
+
+    def local_relative_paths(self) -> Iterator[str]:
+        """
+        Map path to the absolute file system paths, checking if paths exist.
+        Includes single files and files in each resource directory.
+
+        :return: iterator of absolute file paths.
+        """
+        file_path = os.path.relpath(self.path)
+        if not os.path.exists(file_path):
+            raise FileNotFoundError(
+                f"resource file or directory: {self.path} in {os.getcwd()}"
+            )
+        if os.path.isfile(file_path):
+            yield file_path
+        if os.path.isdir(file_path):
+            for root, _, files in os.walk(file_path):
+                for f in files:
+                    dir_file_path = os.path.join(root, f)
+                    yield os.path.relpath(dir_file_path)
+
+
+class FunctionDefinition:
+    def __init__(
+        self,
+        func: Callable[..., Any],
+        project_name: str,
+        account_name: str,
+        asset_type: AssetType,
+        asset_name: str,
+        fabric: Fabric,
+        asset_dependencies: List[AssetPath],
+        pip_dependencies: List[str],
+        resource_paths: List[ResourcePath],
+        assertions: List[Assertion],
+        version_id: Optional[uuid.UUID] = None,
+        repository_id: Optional[uuid.UUID] = None,
+        description: str = "",
+        uri: str = "",
+    ) -> None:
+        self.func = func
+        self.func_name: str = func.__name__
+
+        self.project_name = project_name
+        self.account_name = account_name
+        self.asset_type = asset_type
+        self.asset_name = asset_name
+        self.fabric = fabric
+        self.asset_dependencies = asset_dependencies
+        self.pip_dependencies = pip_dependencies
+        self.resource_paths = resource_paths
+        self.assertions = assertions
+
+        self.version_id = version_id
+        self.repository_id = repository_id
+        self.description = description
+        self.uri = uri
+
+        self.func_source = inspect.getsource(self.func)
+
+        self.source_code_digest = hashlib.sha256()
+        self.source_code_digest.update(self.func_source.encode("utf-8"))
```

Comment:
Something for the future, perhaps: I think this is only relevant for models, but not for datasets, right?

Reply:
Yes, I think so, unless it's a "derived" dataset and we build it ourselves.